### ISSUE TITLE
Subscribing to Subprojects

### ISSUE THREAD

**Issue by bryonbean** [Kitware/CDash#719]
@zackgalbreath @BetsyMcPhail

https://github.com/Kitware/CDash/blob/05b9980e16b5c27e5aed9f34b12f28411be0d9d0/xml_handlers/build_handler.php#L345-L354

Considered with the following fictional build.xml example:
```xml
<Site ...>
  <Subproject name="SubprojectX">
    <Label>SubprojectX</Label>
  </Subproject>
  ...
  <Failure type='warning'>
    ...
    <Labels>
      <Label>SubprojectX</Label>
    </Labels>
  </Failure>
...
</Site>
```

my presumption is that if a user wished to subscribe to build warnings from the sub-project `SubprojectX` that a user could do so by simply ensuring that the Warnings checkbox is checked (on the subscriptions page) then subscribing to the `SubprojectX` label. However the code snippet will prevent a label from being created because it is associated with the subproject. This makes me question my initial presumption that a user should be able to subscribe to a subproject. If there is no label created to associate, and there is no mechanism to subscribe to a subproject then it would not currently be possible—without creating an additional label—to subscribe to this BuildWarning via its sub-project. In the form of a question: _should it be possible to subscribe to a subproject and if so what are the side-effects of adding the label to the BuildError or BuildFailure_?

**Comment by bryonbean** [comment]
Let me rephrase the last part of that question: are there any immediate adverse side-effects of having a label that matches the subproject?

**Comment by zackgalbreath** [comment]
> are there any obvious adverse side-effects of having a label that matches the subproject?

Nope, in fact that's how things currently work.  If you submit the fictional Build.xml file above, a label should get created for `SubprojectX`. 

I would love to decouple labels from subprojects, but that's outside the scope of what you're currently working on.

**Comment by bryonbean** [comment]
FYI, all tests pass* with the following change:

```php
foreach ($this->SubProjects as $subproject => $labels) {
                if (in_array($data, $labels)) {
                    $this->ErrorSubProjectName = $subproject;
                    break;
                }
            }
            $this->Label->SetText($data);
```
I'm going to run with that unless instructed otherwise.

[1]  I currently have 3 javascript tests that fail regularly due to the recalcitrant `angular could not be found on the window`

**Comment by bryonbean** [comment]
> a label should get created for SubprojectX

Associated with the BuildWarning?

**Comment by zackgalbreath** [comment]
> Associated with the BuildWarning?

I tested this out locally.  It does create the label, and the build (both parent and child builds) get associated with that label.  For the purpose of how `sendemail.php` currently works, that should be sufficient for subscribing to a SubProject.

There is also the `label2buildfailure` table which might not be getting populated as much as it should, but I don't think it is used much beyond displaying extra labels on `viewBuildError.php`.

**Comment by zackgalbreath** [comment]
> do you know where labels get set for subprojects?

It's a bit roundabout.  The build_handler calls `$build->SetSubProject(...)` which manages the corresponding label.

**Comment by bryonbean** [comment]
> There is also the label2buildfailure table which might not be getting populated as much as it should.

This actually seems to be the problem, and with very specific test data. If you run `test_multiplesubprojects` you'll see that labels do not get inserted for the BuildFailures. However, here, the data is slightly different, e.g.:
```xml
<Site ...>
  <Subproject name="MyThirdPartyDependency">
    <Label>MyThirdPartyDependency1</Label>
    <Label>MyThirdPartyDependency2</Label>
  </Subproject>
  ...
  <Failure type='Error'>
    ...
    <Labels>
      <Label>MyThirdPartyDependency1</Label>
    </Labels>
  </Failure>
...
</Site>
```

So the question now becomes: _should a user subscribed to receive email for the label `MyThirdPartyDependency1` receive an email regarding this BuildFailure_?

**Comment by BetsyMcPhail** [comment]
> It's a bit roundabout. The build_handler calls $build->SetSubProject(...) which manages the corresponding label.

Okay, thanks. My mental model of that relationship was not correct.

@bryonbean, the test xml files are incorrect. Not only should there only be one label per subproject, like we discussed, but the label and subproject must have the same name.

**Comment by bryonbean** [comment]
> you'll see that labels do not get inserted for the BuildFailures
Let me rephrase: you'll see that labels do not get associated with BuildFailures.

**Comment by bryonbean** [comment]
Thanks @BetsyMcPhail, that solves my problem.
